### PR TITLE
Adds ulimit to the docker module

### PR DESCRIPTION
**Pull request by dusdanig**
The ulimit will be specified as a list and separated by colons. The
hard limit is optional, in which case it is equal to the soft limit.

The ulimits are compared to the ulimits of the container and added
or adjusted accordingly on by a reload.

### CODE REVIEW

**Comment by dusdanig** [comment]
This fixes issue #2465

**Comment by gregdek** [comment]
Thanks @dusdanig. @cove @joshuaconner @softzilla @smashwilson @zfil @ThomasSteinbach please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

File: cloud/docker/docker.py [hunk 1]
```diff
@@ -1566,6 +1596,7 @@ def create_containers(self, count=1):
         else:
             params['host_config']['Memory'] = mem_limit
 
+        params['host_config']['ulimits'] = self.ulimits
```
Comment (smashwilson) [inline cloud/docker/docker.py]: `ulimit` support was added to [docker-py in 1.2.0](https://github.com/docker/docker-py/blob/60d814d742067f3e349119809935f8c0cb6c3b3b/docs/change_log.md#120) and [docker API version 1.18](https://docs.docker.com/engine/reference/api/docker_remote_api/#v1-18-api-changes), so we should gate ulimit support on an `ensure_capabilities` call. You can do this by adding an entry to [`_cap_ver_req`](https://github.com/dusdanig/ansible-modules-core/blob/devel/cloud/docker/docker.py#L654-L669) and then gating this call:

``` python
if self.ulimits is not None:
  self.ensure_capability('ulimits')
  params['host_config']['ulimits'] = self.ulimits
```

File: cloud/docker/docker.py [hunk 2]
```diff
@@ -888,7 +908,7 @@ def get_start_params(self):
         optionals = {}
         for optional_param in ('devices', 'dns', 'volumes_from',
                 'restart_policy', 'restart_policy_retry', 'pid', 'extra_hosts',
-                'log_driver', 'cap_add', 'cap_drop', 'read_only', 'log_opt'):
+                'log_driver', 'cap_add', 'cap_drop', 'read_only', 'log_opt', 'ulimits'):
```
Comment (smashwilson) [inline cloud/docker/docker.py]: Reading through the source, I don't think this is actually used! It adds `ulimits` to `optionals`, but then it never makes its way into `params`.

File: cloud/docker/docker.py [hunk 3]
```diff
@@ -370,6 +370,12 @@
     required: false
     default: 0
     version_added: "2.1"
+  ulimits:
+    description:
+      - ulimits, list ulimits with name, soft and optionally hard limit separated by colons. e.g. nofile:1024:2048
```
Comment (smashwilson) [inline cloud/docker/docker.py]: Would you mind noting the minimum docker-py and docker versions here?

> Requires docker-py >= 1.3.0 and docker >= 1.6.0

**Comment by smashwilson** [comment]
Thanks for tackling this! This'll be handy.

Just a few comments, mostly on ensuring we fail correctly on older versions. It's looking good though.

needs_revision

**Comment by dusdanig** [comment]
Thanks for the review. I really needed this feature ;)

I have updated the code and documentation accordingly.

**Comment by smashwilson** [comment]
:+1: :sparkles: Perfect, thanks.

shipit

**Comment by robynbergeron** [comment]
Thanks again to @dusdanig for this PR, and thanks @cove @joshuaconner @softzilla @smashwilson @zfil @ThomasSteinbach for reviewing. Marking for inclusion.

**Pull request merged**